borsh/tests/test_generic_struct.rs

```diff
@@ -31,6 +41,98 @@ enum B<F, G> {
     Y(G),
 }
 
+#[derive(BorshSerialize, BorshDeserialize, Debug)]
+struct TupleA<T>(T, u32);
+
+#[derive(BorshSerialize, BorshDeserialize, Debug)]
+struct NamedA<T> {
+    a: T,
+    b: u32,
+}
+
+/// `T: PartialOrd` bound is required for `BorshSerialize` derive to be successful
+/// `T: Hash + Eq` bound is required for `BorshDeserialize` derive to be successful
+#[cfg(hash_collections)]
+#[derive(BorshSerialize, BorshDeserialize)]
+struct C<T: PartialOrd + Hash + Eq, U> {
+    a: String,
+    b: HashMap<T, U>,
+}
+
+/// `T: Ord` bound is required for `BorshDeserialize` derive to be successful
+#[derive(BorshSerialize, BorshDeserialize)]
+struct D<T: Ord, U> {
+    a: String,
+    b: BTreeMap<T, U>,
+}
+
+#[cfg(hash_collections)]
+#[derive(BorshSerialize)]
+struct G<K, V, U>(#[borsh_skip] HashMap<K, V>, U);
+
+#[cfg(hash_collections)]
+#[derive(BorshDeserialize)]
+struct G1<K, V, U>(#[borsh_skip] HashMap<K, V>, U);
+
+#[cfg(hash_collections)]
+#[derive(BorshDeserialize)]
+struct G2<K: PartialOrd + Hash + Eq, V, U>(HashMap<K, V>, #[borsh_skip] U);
+
+#[cfg(hash_collections)]
+#[derive(BorshSerialize, BorshDeserialize)]
+struct H<K: Ord, V, U> {
+    x: BTreeMap<K, V>,
+    #[allow(unused)]
+    #[borsh_skip]
+    y: U,
+}
+
+/// `T: Ord` bound is required for `BorshDeserialize` derive to be successful
+#[derive(BorshSerialize, BorshDeserialize, PartialEq, Debug)]
+enum E<T: Ord, U, G> {
+    X { f: BTreeMap<T, U> },
+    Y(G),
+}
+
+#[cfg(hash_collections)]
+#[derive(BorshSerialize, BorshDeserialize, Debug)]
+enum I1<K, V, U> {
+    B {
+        #[allow(unused)]
+        #[borsh_skip]
+        x: HashMap<K, V>,
+        y: String,
+    },
+    C(K, Vec<U>),
+}
+
+#[cfg(hash_collections)]
+#[derive(BorshSerialize, BorshDeserialize, Debug)]
+enum I2<K: PartialOrd + Eq + Hash, V, U> {
+    B { x: HashMap<K, V>, y: String },
+    C(K, #[borsh_skip] U),
+}
+
+trait TraitName {
+    type Associated;
+    fn method(&self);
+}
+
+impl TraitName for u32 {
+    type Associated = String;
+    fn method(&self) {}
+}
+
+#[allow(unused)]
+#[derive(BorshSerialize, BorshDeserialize, PartialEq, Debug)]
+struct Parametrized<T, V>
+where
+    T: TraitName,
+{
+    field: T::Associated,
+    another: V,
+}
+
```

Comment:
serde's `FindTyParams` code isn't perfect.
In commented example it does the right thing.
In a  case, which is synonymous:
```rust
struct Parametrized<T, V> where T: TraitName {

    field: <T as TraitName>::Associated,
    another: V,     
}
```
it produces following expansion
```rust
impl<T, V> borsh::ser::BorshSerialize for Parametrized<T, V>
where
    T: TraitName,
    T: borsh::ser::BorshSerialize,
    V: borsh::ser::BorshSerialize,
{
...
```
which is an error
```bash
error[E0277]: the trait bound `<T as TraitName>::Associated: BorshSerialize` is not satisfied
  --> src/main.rs:28:10
   |
28 | #[derive(BorshSerialize)]
   |          ^^^^^^^^^^^^^^ the trait `BorshSerialize` is not implemented for `<T as TraitName>::Associated`
   |
   = note: this error originates in the derive macro `BorshSerialize` (in Nightly builds, run with -Z macro-backtrace for more info)
                                                                                                                                                                                                                                                                            
For more information about this error, try `rustc --explain E0277`.
```